pkg/csi/controller_server.go

```diff
@@ -0,0 +1,416 @@
+			PersistentVolumeClaim: &corev1.PersistentVolumeClaimVolumeSource{
+				ClaimName: req.VolumeId,
+			},
+		},
+	}
+
+	if err := cs.virtSubresourceRestClient.VirtualMachine(cs.namespace).AddVolume(req.GetNodeId(), opts); err != nil {
+		return nil, status.Errorf(codes.Internal, "Failed to add volume to node %v: %v", req.GetNodeId(), err)
+	}
+
+	checkPVCBound := func(vol *corev1.PersistentVolumeClaim) bool {
+		return vol.Status.Phase == corev1.ClaimBound
+	}
+
+	if !cs.waitForPVCState(req.VolumeId, "Bound", checkPVCBound) {
+		return nil, status.Errorf(codes.DeadlineExceeded, "Failed to attach volume %s to node %s", req.GetVolumeId(), req.GetNodeId())
+	}
+
+	return &csi.ControllerPublishVolumeResponse{}, nil
+}
+
+// ControllerUnpublishVolume will detach the volume
+func (cs *ControllerServer) ControllerUnpublishVolume(ctx context.Context, req *csi.ControllerUnpublishVolumeRequest) (*csi.ControllerUnpublishVolumeResponse, error) {
+	logrus.Infof("ControllerServer ControllerUnpublishVolume req: %v", req)
+
+	if req.GetVolumeId() == "" {
+		return nil, status.Error(codes.InvalidArgument, "Missing volume ID in request")
+	}
+
+	pvc, err := cs.coreClient.PersistentVolumeClaim().Get(cs.namespace, req.GetVolumeId(), metav1.GetOptions{})
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Failed to get PVC %s: %v", req.GetVolumeId(), err)
+	}
+
+	if pvc.Status.Phase != corev1.ClaimBound {
+		return nil, status.Errorf(codes.Aborted, "The PVC %s is in phase %v",
+			req.GetVolumeId(), pvc.Status.Phase)
+	}
+
+	opts := &kv1.RemoveVolumeOptions{
+		Name: req.VolumeId,
+	}
+	if err := cs.virtSubresourceRestClient.VirtualMachine(cs.namespace).RemoveVolume(req.GetNodeId(), opts); err != nil {
+		return nil, status.Errorf(codes.Internal, "Failed to remove volume %v from node %s: %v", req.VolumeId, req.GetNodeId(), err)
+	}
+
+	return &csi.ControllerUnpublishVolumeResponse{}, nil
+}
+
+func (cs *ControllerServer) ListVolumes(context.Context, *csi.ListVolumesRequest) (*csi.ListVolumesResponse, error) {
+	return nil, status.Error(codes.Unimplemented, "")
+}
+
+func (cs *ControllerServer) GetCapacity(context.Context, *csi.GetCapacityRequest) (*csi.GetCapacityResponse, error) {
+	return nil, status.Error(codes.Unimplemented, "")
+}
+
+func (cs *ControllerServer) CreateSnapshot(ctx context.Context, req *csi.CreateSnapshotRequest) (*csi.CreateSnapshotResponse, error) {
+	return nil, status.Error(codes.Unimplemented, "")
+}
+
+func (cs *ControllerServer) DeleteSnapshot(ctx context.Context, req *csi.DeleteSnapshotRequest) (*csi.DeleteSnapshotResponse, error) {
+	return nil, status.Error(codes.Unimplemented, "")
+}
+
+func (cs *ControllerServer) ListSnapshots(context.Context, *csi.ListSnapshotsRequest) (*csi.ListSnapshotsResponse, error) {
+	return nil, status.Error(codes.Unimplemented, "")
+}
+
+func (cs *ControllerServer) ControllerExpandVolume(ctx context.Context, req *csi.ControllerExpandVolumeRequest) (*csi.ControllerExpandVolumeResponse, error) {
```

Comment:
IIRC, longhorn only supports volume expansion when it is currently not in use, how do we manage the validation here?